lib/mpool/src/mclass.c

```diff
@@ -434,28 +434,31 @@ mclass_props_get(struct media_class *const mc, struct mpool_mclass_props *const
     strlcpy(props->mc_path, mc->upath, sizeof(props->mc_path));
 }
 
-merr_t
-mclass_ftw(struct media_class *mc, const char *prefix, struct mpool_file_cb *cb)
+static thread_local struct mpool_file_cb *mclass_ftw_cb;
+static thread_local const char *mclass_ftw_prefix;
+
+static int
+mclass_file_cb(const char *path, const struct stat *sb, int typeflag, struct FTW *ftwbuf)
 {
-    assert(mc);
+    if (typeflag == FTW_D && ftwbuf->level > 0)
+        return FTW_SKIP_SUBTREE;
 
-    /* [HSE_REVISIT]: No nested functions please :). GCC only and my editor
-     * hates it.
-     */
-    int
-    mclass_file_cb(const char *path, const struct stat *sb, int typeflag, struct FTW *ftwbuf)
-    {
-        if (typeflag == FTW_D && ftwbuf->level > 0)
-            return FTW_SKIP_SUBTREE;
+    if (typeflag == FTW_D)
+        return FTW_CONTINUE;
 
-        if (typeflag == FTW_D)
-            return FTW_CONTINUE;
+    if (!mclass_ftw_prefix || strstr(basename(path), mclass_ftw_prefix))
+        mclass_ftw_cb->cbfunc(mclass_ftw_cb->cbarg, path);
 
-        if (!prefix || strstr(basename(path), prefix))
```

Comment:
nftw is great for directory traversal, but I wish its callback function accepted a user-provided callback pointer :-(

Reply:
IKR?  maybe fts_open()... ?

Reply:
A quick glance through the fts_open() man page doesn't indicate that it accepts a user-provided cb pointer.

Reply:
fts_pointer, but it's not clear how to set it up for the first iteration or whether it's carried from call to call (I've never used this interface).

Reply:
Ah, I missed that field in FTSENT, thanks! However, it is not clear from the man page on how to initialize fts_number/fts_pointer. You just introduced me to fts, never heard about it before :-)

In any case, the nice thing about fts functions is that we explicitly call fts_read() to traverse the file hierarchy. So, we have the flexibility to do whatever with FTSENT (or each file) returned by fts_read(), unlike nftw() where a user supplied cb is invoked for each file but that fails to accept a user context.

I will try to reimplement some of the nftw() usage in mpool with fts functions. I can get rid of some of the static globals that way...